numbat/src/command.rs

```diff
@@ -46,49 +53,365 @@ impl FromStr for CommandKind {
     }
 }
 
-#[derive(Debug, Clone, PartialEq)]
-pub enum Command<'a> {
-    Help,
-    Info { item: &'a str },
-    List { items: Option<ListItems> },
-    Clear,
-    Save { dst: &'a str },
+#[derive(Debug, PartialEq, Eq)]
+#[must_use]
+pub enum CommandControlFlow {
+    Continue,
+    Return,
+    NotACommand,
+}
+
+pub struct CommandContext<'ctx, 'editor, Editor> {
+    pub ctx: &'ctx mut Context,
+    pub editor: &'editor mut Editor,
+}
+
+#[derive(Debug)]
+pub enum CommandError {
+    Parse(ParseError),
+    Runtime(RuntimeError),
+}
+
+impl From<ParseError> for CommandError {
+    fn from(err: ParseError) -> Self {
+        Self::Parse(err)
+    }
+}
+
+impl From<RuntimeError> for CommandError {
+    fn from(err: RuntimeError) -> Self {
+        Self::Runtime(err)
+    }
+}
+
+impl ErrorDiagnostic for CommandError {
+    fn diagnostics(&self) -> Vec<crate::Diagnostic> {
+        match self {
+            CommandError::Parse(parse_error) => parse_error.diagnostics(),
+            CommandError::Runtime(runtime_error) => runtime_error.diagnostics(),
+        }
+    }
+}
+
+enum Command<'session, 'input, Editor> {
+    Help {
+        print_fn: fn(&Markup),
+    },
+    Info {
+        item: &'input str,
+        print_fn: fn(&Markup),
+    },
+    List {
+        items: Option<ListItems>,
+        print_fn: fn(&Markup),
+    },
+    Clear {
+        clear_fn: fn(&mut Editor) -> CommandControlFlow,
+    },
+    Save(SaveCmdArgs<'session, 'input>),
     Quit,
 }
 
-/// Contains just the words and word boundaries of the input we're parsing, no
-/// `code_source_id`
+struct SaveCmdArgs<'session, 'input> {
+    session_history: &'session SessionHistory,
+    dst: &'input str,
+    print_fn: Option<fn(&Markup)>,
+}
+
+impl SaveCmdArgs<'_, '_> {
+    fn save(&self) -> Result<(), Box<RuntimeError>> {
+        let Self {
+            session_history,
+            dst,
+            print_fn,
+        } = self;
+
+        session_history.save(
+            dst,
+            SessionHistoryOptions {
+                include_err_lines: false,
+                trim_lines: true,
+            },
+        )?;
+
+        if let Some(print_markup) = print_fn {
+            let markup = m::text("successfully saved session history to")
+                + m::space()
+                + m::string(dst.to_compact_string());
+            print_markup(&markup)
+        }
+
+        Ok(())
+    }
+}
+
+pub struct CommandRunner<Editor> {
```

Comment:
Minor: Maybe use
```suggestion
pub struct CommandRunner<Editor = ()> {
```
here, so clients which do not set up a `clear` command can simply use `CommandRunner` instead of `CommandRunner::<()>`